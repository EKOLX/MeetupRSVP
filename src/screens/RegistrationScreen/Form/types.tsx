import { InputsModel } from "../types";

export interface Props {
  values: InputsModel;
  errors: Partial<InputsModel>;
  onInputChange: (text: string, input: keyof InputsModel) => void;
  onError: (errorMessage: string | null, input: keyof InputsModel) => void;
  onSubmit: () => void;
}
