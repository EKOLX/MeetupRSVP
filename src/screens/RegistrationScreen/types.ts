export interface InputsModel {
  name: string;
  age: string;
  birthdate: string;
  profession: string;
  locality: string;
  guestsCount: string;
  address: string;
}

export interface Validation {
  valid: boolean;
  inputs: InputsModel;
}
