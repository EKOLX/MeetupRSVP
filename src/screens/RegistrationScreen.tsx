import { StyleSheet, Text, View } from "react-native";

const RegistrationScreen = () => (
  <View style={styles.container}>
    <Text>Registration Screen</Text>
  </View>
);

export default RegistrationScreen;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center",
  },
});
