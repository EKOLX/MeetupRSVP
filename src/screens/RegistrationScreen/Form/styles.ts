import { StyleSheet } from "react-native";

export const styles = StyleSheet.create({
  container: {
    flex: 1,
    width: "100%",
  },
  multiline: {
    height: 81,
  },
  submit: {
    marginTop: 32,
  },
});
