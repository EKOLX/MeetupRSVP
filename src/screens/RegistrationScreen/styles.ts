import { StyleSheet } from "react-native";

export const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center",
    backgroundColor: "#FFFFFF",
    padding: 32,
  },
  header: {
    fontSize: 22,
    fontWeight: "600",
    textAlign: "center",
    marginBottom: 32,
  },
});
