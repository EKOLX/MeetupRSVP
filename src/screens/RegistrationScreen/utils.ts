import { InputsModel, Validation } from "./types";

export const validateInputs = (inputs: InputsModel): Validation => {
  const validation: Validation = { valid: true, inputs: {} as InputsModel };

  if (!inputs.name) {
    validation.inputs.name = "Please input your name";
  }

  if (!inputs.age) {
    validation.inputs.age = "Please input your age";
  } else if (
    !/^\d+$/.test(inputs.age) ||
    isNaN(parseInt(inputs.age)) ||
    parseInt(inputs.age) < 4
  ) {
    validation.inputs.age = "Please input correct age";
  }

  if (!inputs.birthdate) {
    validation.inputs.birthdate = "Please input your date of birth";
  } else if (!parseDate(inputs.birthdate)) {
    validation.inputs.birthdate = "Please input correct date of birth";
  }

  if (!inputs.profession) {
    validation.inputs.profession = "Please input your profession";
  } else if (
    inputs.profession !== "Employed" &&
    inputs.profession !== "Student"
  ) {
    validation.inputs.profession =
      "Only two types of professions are supported: 'Employed' or 'Student'";
  }

  if (!inputs.locality) {
    validation.inputs.locality = "Please input your locality";
  }

  if (!inputs.guestsCount) {
    validation.inputs.guestsCount = "Please input number of guests";
  } else if (!/^0|1|2$/.test(inputs.guestsCount)) {
    validation.inputs.guestsCount = "Allowed number of guests: 0, 1 or 2";
  }

  if (!inputs.address) {
    validation.inputs.address = "Please input your address";
  }

  // If any input has error message make validation as invalid
  for (const input in validation.inputs) {
    if (validation.inputs[input as keyof InputsModel]) {
      validation.valid = false;
      break;
    }
  }

  return validation;
};

function parseDate(value: string) {
  const validDate = value.match(/^(\d{1,2}).(\d{1,2}).(\d{4})$/);
  return validDate
    ? new Date(
        parseInt(validDate[3]),
        parseInt(validDate[2]) - 1,
        parseInt(validDate[1])
      )
    : null;
}
