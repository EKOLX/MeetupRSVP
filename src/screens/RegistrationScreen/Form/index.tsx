import { FC } from "react";
import { KeyboardAvoidingView, ScrollView } from "react-native";

import Input from "../../../components/UI/Input";
import Button from "../../../components/UI/Button";
import { Props } from "./types";
import { styles } from "./styles";

const Form: FC<Props> = ({ errors, onInputChange, onError, onSubmit }) => {
  return (
    <KeyboardAvoidingView style={styles.container} behavior="padding">
      <ScrollView showsVerticalScrollIndicator={false}>
        <Input
          label="Name"
          placeholder="Jack"
          error={errors.name}
          onFocus={() => onError(null, "name")}
          onChangeText={(text) => onInputChange(text, "name")}
        />
        <Input
          label="Age"
          placeholder="33"
          inputMode="numeric"
          error={errors.age}
          onFocus={() => onError(null, "age")}
          onChangeText={(text) => onInputChange(text, "age")}
        />
        <Input
          label="Date of Birth"
          placeholder="22.06.1988"
          inputMode="decimal"
          error={errors.birthdate}
          onFocus={() => onError(null, "birthdate")}
          onChangeText={(text) => onInputChange(text, "birthdate")}
        />
        <Input
          label="Profession"
          placeholder="Employed | Student"
          error={errors.profession}
          onFocus={() => onError(null, "profession")}
          onChangeText={(text) => onInputChange(text, "profession")}
        />
        <Input
          label="Locality"
          placeholder="Dubai"
          error={errors.locality}
          onFocus={() => onError(null, "locality")}
          onChangeText={(text) => onInputChange(text, "locality")}
        />
        <Input
          label="Number of Guests"
          placeholder="0-2"
          inputMode="numeric"
          error={errors.guestsCount}
          onFocus={() => onError(null, "guestsCount")}
          onChangeText={(text) => onInputChange(text, "guestsCount")}
        />
        <Input
          style={styles.multiline}
          label="Address"
          placeholder="Dzhorat street"
          multiline
          numberOfLines={4}
          maxLength={50}
          textAlignVertical="top"
          blurOnSubmit
          error={errors.address}
          onFocus={() => onError(null, "address")}
          onChangeText={(text) => onInputChange(text, "address")}
        />
      </ScrollView>

      <Button style={styles.submit} title="Submit" onPress={onSubmit} />
    </KeyboardAvoidingView>
  );
};

export default Form;
