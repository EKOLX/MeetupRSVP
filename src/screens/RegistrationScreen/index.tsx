import { useState } from "react";
import { Alert, Keyboard, Text, View } from "react-native";

import Form from "./Form";
import Loader from "../../components/UI/Loader";
import { validateInputs } from "./utils";
import { RootDrawerScreenProps } from "../../navigation/types";
import { InputsModel } from "./types";
import { styles } from "./styles";

const RegistrationScreen = ({
  navigation,
}: RootDrawerScreenProps<"Registration">) => {
  const [loading, setLoading] = useState(false);
  const [inputs, setInputs] = useState<InputsModel>({
    name: "",
    age: "",
    birthdate: "",
    profession: "",
    locality: "",
    guestsCount: "",
    address: "",
  });
  const [inputErrors, setInputErrors] = useState<Partial<InputsModel>>({});

  const handleBeforeLeavingScreen = (event: any) => {
    let hasUnsavedChanges = false;

    for (const input in inputs) {
      if (inputs[input as keyof InputsModel]) {
        hasUnsavedChanges = true;
        break;
      }
    }

    if (hasUnsavedChanges) {
      Alert.alert(
        "Discard changes?",
        "You have unsaved changes. Are you sure to discard them and leave the screen?",
        [
          { text: "Don't leave", style: "cancel" },
          {
            text: "Discard",
            style: "destructive",
            onPress: () => {
              // TODO: Leave the screen
            },
          },
        ]
      );
    }
  };

  const handleOnInputChange = (text: string, input: keyof InputsModel) => {
    setInputs((prevState) => ({ ...prevState, [input]: text }));
  };

  const handleOnError = (
    errorMessage: string | null,
    input: keyof InputsModel
  ) => {
    setInputErrors((prevState) => ({ ...prevState, [input]: errorMessage }));
  };

  const validate = (): boolean => {
    let { valid, inputs: validatedInputs } = validateInputs(inputs);

    if (!valid) {
      for (const input in validatedInputs) {
        const key = input as keyof InputsModel;
        if (validatedInputs[key]) handleOnError(validatedInputs[key], key);
      }
    }

    return valid;
  };

  const handleOnSubmit = () => {
    Keyboard.dismiss();

    if (validate()) {
      setLoading(true);
      // TODO: register
      setTimeout(() => {
        setLoading(false);
        navigation.navigate("Users");
      }, 3300);
    }
  };

  return (
    <View style={styles.container}>
      <Text style={styles.header}>
        Start registration to participate in the meetup
      </Text>

      <Form
        errors={inputErrors}
        onInputChange={handleOnInputChange}
        onError={handleOnError}
        onSubmit={handleOnSubmit}
      />

      <Loader visible={loading} />
    </View>
  );
};

export default RegistrationScreen;
