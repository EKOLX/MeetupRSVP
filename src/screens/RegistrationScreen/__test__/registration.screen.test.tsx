import { Text, TextInput } from "react-native";
import { create, act, ReactTestRenderer } from "react-test-renderer";

import RegistrationScreen from "../index";
import Messages from "../../../constants/Messages";
import { RootDrawerScreenProps } from "../../../navigation/types";
import Form from "../../../components/Form";

jest.runAllTimers();
jest.mock("react-redux", () => {
  const ActualReactRedux = jest.requireActual("react-redux");
  return {
    ...ActualReactRedux,
    useDispatch: jest.fn(),
    useSelector: jest.fn().mockImplementation(() => ({})),
  };
});

describe("Registration screen", () => {
  let registrationScreen: ReactTestRenderer;

  beforeEach(() => {
    registrationScreen = create(
      <RegistrationScreen {...({} as RootDrawerScreenProps<"Registration">)} />
    );
  });

  it("should match snapshot", () => {
    const tree = registrationScreen.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should show error messages if Form input fields haven't been filled", () => {
    const buttonSubmit = registrationScreen.root.findByProps({
      testID: "buttonSubmit",
    });
    act(() => buttonSubmit.props.onPress());

    const inputNameTexts = registrationScreen.root
      .findByProps({ testID: "inputName" })
      .findAllByType(Text as any);
    const inputNameErrorText = inputNameTexts[1].props;

    const inputAgeTexts = registrationScreen.root
      .findByProps({ testID: "inputAge" })
      .findAllByType(Text as any);
    const inputAgeErrorText = inputAgeTexts[1].props;

    const inputBirthdateTexts = registrationScreen.root
      .findByProps({ testID: "inputBirthdate" })
      .findAllByType(Text as any);
    const inputBirthdateErrorText = inputBirthdateTexts[1].props;

    const inputProfessionTexts = registrationScreen.root
      .findByProps({ testID: "inputProfession" })
      .findAllByType(Text as any);
    const inputProfessionErrorText = inputProfessionTexts[1].props;

    const inputLocalityTexts = registrationScreen.root
      .findByProps({ testID: "inputLocality" })
      .findAllByType(Text as any);
    const inputLocalityErrorText = inputLocalityTexts[1].props;

    const inputGuestsCountTexts = registrationScreen.root
      .findByProps({ testID: "inputGuestsCount" })
      .findAllByType(Text as any);
    const inputGuestsCountErrorText = inputGuestsCountTexts[1].props;

    const inputAddressTexts = registrationScreen.root
      .findByProps({ testID: "inputAddress" })
      .findAllByType(Text as any);
    const inputAddressErrorText = inputAddressTexts[1].props;

    expect(inputNameTexts.length).toEqual(2);
    expect(inputNameErrorText.children).toEqual(
      Messages.please_input_your_name
    );

    expect(inputAgeTexts.length).toEqual(2);
    expect(inputAgeErrorText.children).toEqual(Messages.please_input_your_age);

    expect(inputBirthdateTexts.length).toEqual(2);
    expect(inputBirthdateErrorText.children).toEqual(
      Messages.please_input_your_date_of_birth
    );

    expect(inputProfessionTexts.length).toEqual(2);
    expect(inputProfessionErrorText.children).toEqual(
      Messages.please_input_your_profession
    );

    expect(inputLocalityTexts.length).toEqual(2);
    expect(inputLocalityErrorText.children).toEqual(
      Messages.please_input_your_locality
    );

    expect(inputGuestsCountTexts.length).toEqual(2);
    expect(inputGuestsCountErrorText.children).toEqual(
      Messages.please_input_number_of_guests
    );

    expect(inputAddressTexts.length).toEqual(2);
    expect(inputAddressErrorText.children).toEqual(
      Messages.please_input_your_address
    );
  });

  it("shouldn't have any errors in Form after inputs filled in", () => {
    const buttonSubmit = registrationScreen.root.findByProps({
      testID: "buttonSubmit",
    });
    const inputName = registrationScreen.root.findByProps({
      testID: "inputName",
    });
    const inputAge = registrationScreen.root.findByProps({
      testID: "inputAge",
    });
    const inputBirthdate = registrationScreen.root.findByProps({
      testID: "inputBirthdate",
    });
    const inputProfession = registrationScreen.root.findByProps({
      testID: "inputProfession",
    });
    const inputLocality = registrationScreen.root.findByProps({
      testID: "inputLocality",
    });
    const inputGuestsCount = registrationScreen.root.findByProps({
      testID: "inputGuestsCount",
    });
    const inputAddress = registrationScreen.root.findByProps({
      testID: "inputAddress",
    });

    act(() => {
      inputName.props.onChangeText("EL Khan");
      inputAge.props.onChangeText("33");
      inputBirthdate.props.onChangeText("22.06.2002");
      inputProfession.props.onChangeText("Employed");
      inputLocality.props.onChangeText("Prague");
      inputGuestsCount.props.onChangeText("1");
      inputAddress.props.onChangeText("Elm Street");
    });

    const form = registrationScreen.root.findByType(Form);

    act(() => buttonSubmit.props.onPress());

    expect(form.props.errors).toEqual({});
  });
});
