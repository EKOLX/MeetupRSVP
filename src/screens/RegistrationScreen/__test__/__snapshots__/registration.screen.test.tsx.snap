// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registration screen should match snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "center",
      "padding": 32,
    }
  }
>
  <Text
    style={
      {
        "fontSize": 22,
        "fontWeight": "600",
        "marginBottom": 32,
        "textAlign": "center",
      }
    }
  >
    Start registration to participate in the meetup
  </Text>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "flex": 1,
          "width": "100%",
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      showsVerticalScrollIndicator={false}
    >
      <View>
        <View
          style={
            {
              "marginVertical": 6,
              "width": "100%",
            }
          }
          testID="inputName"
        >
          <Text
            style={
              {
                "color": "#BACDDB",
                "fontSize": 13,
                "marginBottom": 3,
              }
            }
          >
            Name
          </Text>
          <TextInput
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="Jack"
            placeholderTextColor="#BACDDB"
            returnKeyType="done"
            style={
              [
                {
                  "borderRadius": 6,
                  "borderWidth": 1,
                  "fontSize": 16,
                  "padding": 6,
                },
                undefined,
                {
                  "borderColor": "#BACDDB",
                },
              ]
            }
          />
        </View>
        <View
          style={
            {
              "marginVertical": 6,
              "width": "100%",
            }
          }
          testID="inputAge"
        >
          <Text
            style={
              {
                "color": "#BACDDB",
                "fontSize": 13,
                "marginBottom": 3,
              }
            }
          >
            Age
          </Text>
          <TextInput
            inputMode="numeric"
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="33"
            placeholderTextColor="#BACDDB"
            returnKeyType="done"
            style={
              [
                {
                  "borderRadius": 6,
                  "borderWidth": 1,
                  "fontSize": 16,
                  "padding": 6,
                },
                undefined,
                {
                  "borderColor": "#BACDDB",
                },
              ]
            }
          />
        </View>
        <View
          style={
            {
              "marginVertical": 6,
              "width": "100%",
            }
          }
          testID="inputBirthdate"
        >
          <Text
            style={
              {
                "color": "#BACDDB",
                "fontSize": 13,
                "marginBottom": 3,
              }
            }
          >
            Date of Birth
          </Text>
          <TextInput
            inputMode="decimal"
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="22.06.1988"
            placeholderTextColor="#BACDDB"
            returnKeyType="done"
            style={
              [
                {
                  "borderRadius": 6,
                  "borderWidth": 1,
                  "fontSize": 16,
                  "padding": 6,
                },
                undefined,
                {
                  "borderColor": "#BACDDB",
                },
              ]
            }
          />
        </View>
        <View
          style={
            {
              "marginVertical": 6,
              "width": "100%",
            }
          }
          testID="inputProfession"
        >
          <Text
            style={
              {
                "color": "#BACDDB",
                "fontSize": 13,
                "marginBottom": 3,
              }
            }
          >
            Profession
          </Text>
          <TextInput
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="Employed | Student"
            placeholderTextColor="#BACDDB"
            returnKeyType="done"
            style={
              [
                {
                  "borderRadius": 6,
                  "borderWidth": 1,
                  "fontSize": 16,
                  "padding": 6,
                },
                undefined,
                {
                  "borderColor": "#BACDDB",
                },
              ]
            }
          />
        </View>
        <View
          style={
            {
              "marginVertical": 6,
              "width": "100%",
            }
          }
          testID="inputLocality"
        >
          <Text
            style={
              {
                "color": "#BACDDB",
                "fontSize": 13,
                "marginBottom": 3,
              }
            }
          >
            Locality
          </Text>
          <TextInput
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="Dubai"
            placeholderTextColor="#BACDDB"
            returnKeyType="done"
            style={
              [
                {
                  "borderRadius": 6,
                  "borderWidth": 1,
                  "fontSize": 16,
                  "padding": 6,
                },
                undefined,
                {
                  "borderColor": "#BACDDB",
                },
              ]
            }
          />
        </View>
        <View
          style={
            {
              "marginVertical": 6,
              "width": "100%",
            }
          }
          testID="inputGuestsCount"
        >
          <Text
            style={
              {
                "color": "#BACDDB",
                "fontSize": 13,
                "marginBottom": 3,
              }
            }
          >
            Number of Guests
          </Text>
          <TextInput
            inputMode="numeric"
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="0-2"
            placeholderTextColor="#BACDDB"
            returnKeyType="done"
            style={
              [
                {
                  "borderRadius": 6,
                  "borderWidth": 1,
                  "fontSize": 16,
                  "padding": 6,
                },
                undefined,
                {
                  "borderColor": "#BACDDB",
                },
              ]
            }
          />
        </View>
        <View
          style={
            {
              "marginVertical": 6,
              "width": "100%",
            }
          }
          testID="inputAddress"
        >
          <Text
            style={
              {
                "color": "#BACDDB",
                "fontSize": 13,
                "marginBottom": 3,
              }
            }
          >
            Address
          </Text>
          <TextInput
            blurOnSubmit={true}
            maxLength={50}
            multiline={true}
            numberOfLines={2}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="Elm Street"
            placeholderTextColor="#BACDDB"
            returnKeyType="done"
            style={
              [
                {
                  "borderRadius": 6,
                  "borderWidth": 1,
                  "fontSize": 16,
                  "padding": 6,
                },
                {
                  "height": 55,
                },
                {
                  "borderColor": "#BACDDB",
                },
              ]
            }
            textAlignVertical="top"
          />
        </View>
      </View>
    </RCTScrollView>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "backgroundColor": "#6D67E4",
            "borderRadius": 16,
            "paddingVertical": 16,
            "width": "100%",
          },
          {
            "marginTop": 32,
          },
        ]
      }
      testID="buttonSubmit"
    >
      <Text
        style={
          {
            "color": "#FFFFFF",
            "fontSize": 16,
            "fontWeight": "500",
            "textAlign": "center",
          }
        }
      >
        Submit
      </Text>
    </View>
  </View>
</View>
`;
